### ISSUE TITLE
Stack ignores setup-depends altogether

### ISSUE THREAD

**Issue by Blaisorblade** [commercialhaskell/stack#2515]
# Actual behavior

in [this log](http://lpaste.net/3330066059314069504), stack tries building `cairo` before `gtk2hs-buildtools`, and this fails. Luckily it builds `gtk2hs-buildtools`. Restarting the build then works.
# Expected behavior

stack should build `cairo` after `gtk2hs-buildtools`.

The dependency is declared by the cabal file but uses `setup-depends` (at least on HEAD)—this might trigger the bug:

https://github.com/gtk2hs/gtk2hs/blob/3fdb125c57e853a9ea473a54e277600be7d962bd/cairo/cairo.cabal#L46-L49

EDIT: this is split out from #2512.

**Comment by Blaisorblade** [comment]
Other example: http://lpaste.net/6497164652642828288.

**Comment by Blaisorblade** [comment]
The problem is in packageFromPackageDescription.
However, fixing this requires Cabal-1.24. BTW, the build file even writes `Cabal-Version:  >= 1.24`, and stack totally ignores that.

Indeed, we need `setupBuildInfo`, introduced in https://github.com/haskell/cabal/pull/2515 and first supported in https://hackage.haskell.org/package/Cabal-1.24.0.0/changelog.

**Comment by Blaisorblade** [comment]
I just realized we ignore setup-depends altogether, *including* the version requirements. So we might end up not installing `gtk2hs-buildtools`, or installing the wrong version. I tried checking with `stack build --dry-run cairo`, but it uses an older version; the offending one is not even on Hackage yet, only on Github.

**Comment by Blaisorblade** [comment]
Welp. To get this right we might need to have a separate plan for building setup, otherwise we might get regressions. That sounds nontrivial.
A possible regression would happen, for instance, if a package setup file requires, for building, a Cabal version that is different from the one we're picking.
I think the proper solution can involve using a separate `.stack-work` folder for the setup file and for any dependencies it specifies.

I can't decide how often an "imperfect" solution would work, and it it would work enough to be worth implementing, but I suspect yes; and some parts are reusable later.

https://www.well-typed.com/blog/2015/07/cabal-setup-deps/
https://www.well-typed.com/blog/2015/03/qualified-goals/

**Comment by Blaisorblade** [comment]
# To reproduce

Try building `cairo` using this `stack.yaml`:

```yaml
resolver: lts-6.6

packages:
- location:
    git: https://github.com/gtk2hs/gtk2hs
    commit: 3fdb125c57e853a9ea473a54e277600be7d962bd
  subdirs:
  - cairo

extra-deps:
```

Run:
```
$ stack build cairo --dry-run
Cabal file warning in /Users/pgiarrusso/AeroFS/Repos/stack-bluevelvet/test-cairo/.stack-work/downloaded/QJf9nhMltEuF/cairo/cairo.cabal: Ignoring unknown section type: custom-setup
Cabal file warning in /Users/pgiarrusso/AeroFS/Repos/stack-bluevelvet/test-cairo/.stack-work/downloaded/QJf9nhMltEuF/cairo/cairo.cabal: This package requires at least Cabal version 1.24
Cabal file warning in /Users/pgiarrusso/AeroFS/Repos/stack-bluevelvet/test-cairo/.stack-work/downloaded/QJf9nhMltEuF/cairo/cairo.cabal: Ignoring unknown section type: custom-setup
Cabal file warning in /Users/pgiarrusso/AeroFS/Repos/stack-bluevelvet/test-cairo/.stack-work/downloaded/QJf9nhMltEuF/cairo/cairo.cabal: This package requires at least Cabal version 1.24
No packages would be unregistered.

Would build:
cairo-0.13.3.0: database=local, source=/Users/pgiarrusso/AeroFS/Repos/stack-bluevelvet/test-cairo/.stack-work/downloaded/QJf9nhMltEuF/cairo/

No executables to be installed.
```

In comparison, inside the current stack folder (with `lts-6.0`), I get the more plausible

```
$ stack build cairo --dry-run
No packages would be unregistered.

Would build:
alex-3.1.7: database=snapshot, source=package index
cairo-0.13.1.1: database=snapshot, source=package index, after: gtk2hs-buildtools-0.13.0.5
gtk2hs-buildtools-0.13.0.5: database=snapshot, source=package index, after: alex-3.1.7,hashtables-1.2.1.0
hashtables-1.2.1.0: database=snapshot, source=package index

No executables to be installed.
```

**Comment by mgsloan** [comment]
Yeah, this is known.  We just didn't implement it, lack of it has never caused us any difficulty.

https://github.com/commercialhaskell/stack/issues/2094

> Welp. To get this right we might need to have a separate plan for building setup, otherwise we might get regressions. 

Why?  I don't see any issue with using a single set of package versions for both Setup.hs.  Varying those versions is a degree of flexibility that I find unnecessary and overcomplicated.

**Comment by Blaisorblade** [comment]
> I don't see any issue with using a single set of package versions for both Setup.hs.

TL;DR. It will work in only like 99% of the cases, and with Cabal 2.0 it might well break more often, but it *should* work often enough that it's probably worth implementing it at first the "easy way" and tracking the rest as a known limitation.

According to the blog posts (and even in practice), two packages might well have conflicting version requirements on, say, Cabal for setup-depends. The libraries themselves don't depend on Cabal so they're in fact compatible, and the two different Cabal versions are never linked together, but the setup files can't easily coexist in a Stack build plan.
For now most packages with `setup-depends: Cabal` accept Cabal 1.24. With Cabal 2.0 that might change.

The only exception is `base-noprelude-4.7.0.*`, whose setup requires Cabal <=1.21. I'm not sure whether the problematic build plans would build at all today (I'm trying though), but they certainly won't build with the easy implementation, so we *risk* a regression.

**Comment by mgsloan** [comment]
We now have support for the `custom-setup` stanza, so I believe this is resolved!
